community/kernel/src/main/java/org/neo4j/kernel/monitoring/VmPauseMonitor.java

```diff
@@ -0,0 +1,176 @@
+/*
+ * Copyright (c) 2002-2018 "Neo Technology,"
+ * Network Engine for Objects in Lund AB [http://neotechnology.com]
+ *
+ * This file is part of Neo4j.
+ *
+ * Neo4j is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package org.neo4j.kernel.monitoring;
+
+import java.lang.management.GarbageCollectorMXBean;
+import java.lang.management.ManagementFactory;
+import java.util.function.Consumer;
+
+import org.neo4j.logging.Log;
+
+import static java.lang.String.format;
+import static java.lang.System.nanoTime;
+import static java.lang.Thread.currentThread;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+import static java.util.concurrent.TimeUnit.NANOSECONDS;
+import static org.neo4j.util.Preconditions.requirePositive;
+
+public final class VmPauseMonitor
+{
+    private final long measurementDurationnNs;
+    private final long stallAlertThresholdNs;
+    private final Log log;
+    private final Consumer<VmPauseInfo> listener;
+    private final Thread worker;
+
+    public VmPauseMonitor( long measureIntervalMs, long stallAlertThresholdMs, Log log, Consumer<VmPauseInfo> listener )
+    {
+        this.measurementDurationnNs = MILLISECONDS.toNanos( requirePositive( measureIntervalMs ) );
+        this.stallAlertThresholdNs = MILLISECONDS.toNanos( requirePositive( stallAlertThresholdMs ) );
+        this.log = requireNonNull( log );
+        this.listener = requireNonNull( listener );
+        this.worker = new Thread( this::run );
```

Comment:
we have Neo4jJobScheduler that is responsible for all of the threads, so let's use that instead for thread things